### PR TITLE
display quoted notes #944

### PR DESCRIPTION
**Pull request by bryanmontz**
## Issues covered
#944 

## Description
This PR changes the "Link to note" rendering of links to notes in rendered note cards to show the note itself.

## How to test
1. Navigate to any note that refers to another note.
2. View the quoted note, tap on it to load.

## Screenshots/Video
| Before | After |
|--------|--------|
|old|new|

### CODE REVIEW

File: Nos/Extensions/URL+Extensions.swift
Comment (martindsq) [inline Nos/Extensions/URL+Extensions.swift:47]: The computed var is almost self-explanatory, but I think there is still a benefit if we document it anyway so that new developers can understand what they are. Maybe a documentation entry explaining how note links look like and that they are generated by NoteParser would do.

File: Nos/Models/CoreData/Event+CoreDataClass.swift
Comment (martindsq) [inline Nos/Models/CoreData/Event+CoreDataClass.swift:561]: This looks like something that could have side-effects in other parts of the app. Mind if I ping @mplorentz so he takes a look at it?

File: Nos/Models/CoreData/Event+CoreDataClass.swift
Comment (martindsq) [inline Nos/Models/CoreData/Event+CoreDataClass.swift:826]: Given that loadFirstQuotedNote() makes requests to relays, can you refactor this so setting quotedNoteID doesn't have side effects like this? It makes it harder to track what's happening in the app. I'm fine if that means having Task { await loadFirstQuotedNote() } in a couple of places. I think it would make the code easier to understand.

File: Nos/Models/CoreData/Event+CoreDataClass.swift
Comment (martindsq) [inline Nos/Models/CoreData/Event+CoreDataClass.swift:862]: This line of code was only called in the else condition before, I just want to double check having it here is desirable.

**Review by martindsq — DISMISSED** [review]
I'm excited to have this on main soon. Quoted notes are great fine for me (I'm not a big fan of the design though, I hope we polish it a little bit afterwards). I'm leaving some comments. I have testing it even more pending.

File: Nos/Extensions/URL+Extensions.swift
Comment (bryanmontz) [reply Nos/Extensions/URL+Extensions.swift:47]: Yeah, good point. However, I ended up not using this in the end, so I'll just remove it for now.

File: Nos/Models/CoreData/Event+CoreDataClass.swift
Comment (bryanmontz) [reply Nos/Models/CoreData/Event+CoreDataClass.swift:862]: I believe that it is desirable. Otherwise, `loadingViewData` will never get set back to `false`, which could be misleading both to UI that depends on it as well as to devs trying to debug issues in this area. Would be great to get confirmation from @mplorentz if possible though.

File: Nos/Models/CoreData/Event+CoreDataClass.swift
Comment (bryanmontz) [reply Nos/Models/CoreData/Event+CoreDataClass.swift]: Moved it just to one spot.

File: Nos/Models/CoreData/Event+CoreDataClass.swift
Comment (bryanmontz) [reply Nos/Models/CoreData/Event+CoreDataClass.swift]: I moved it to avoid side-effects from other callers of this function.

File: NosTests/Models/NoteParserTests.swift
Comment (bryanmontz) [inline NosTests/Models/NoteParserTests.swift:239]: I wasn't sure what to do with this. It's probably a Product question for Linda what to do when a note link is inline with other content as opposed to just at the end of a note.

File: Nos/Models/NoteParser.swift
Comment (joshuatbrown) [inline Nos/Models/NoteParser.swift:34]: I think it's time to rename this or the other `parse(content:tags:context:)` -- it's too confusing and annoying to have to specify the return type when we call either of them. I know you didn't name it, but would you mind fixing it? I'd suggest `components(...)` like Apple's Calendar API.

```suggestion
    func components(from content: String, tags: [[String]], context: NSManagedObjectContext) -> NoteDisplayComponents {
```

**Review by joshuatbrown — COMMENTED** [review]
A few thoughts for now -- I didn't have time for a full review, but what I saw is looking great!

File: Nos/Models/URLParser.swift
Comment (bryanmontz) [reply Nos/Models/URLParser.swift:6]: I have a small, typos-only PR coming. 😉

File: Nos/Models/NoteParser.swift
Comment (joshuatbrown) [reply Nos/Models/NoteParser.swift:16]: Yeah, I think you're right. I just wanted to think it through. I'm good with keeping it as a struct.

File: Nos/Models/NoteParser.swift
Comment (joshuatbrown) [inline Nos/Models/NoteParser.swift:9]: Nitpick:
```suggestion
    /// Encapsulates parsed components of a note that can be used to display the note in the UI.
```
I always try to avoid using the type in the doc comments. I usually start by writing it exactly how you did here, then I delete the part where I said "This class" or "This struct" or whatever. 😀

File: Nos/Models/NoteParser.swift
Comment (joshuatbrown) [inline Nos/Models/NoteParser.swift:126]: We've been adding default values to doc comments like this:

```suggestion
    ///   - capturesFirstNote: If true, this function will extract the first quoted note id, if it exists. Defaults to `false`.
```

File: Nos/Views/Modifiers/View+StyledBorder.swift
Comment (joshuatbrown) [inline Nos/Views/Modifiers/View+StyledBorder.swift:19]: We use SwiftGen to generate color constants from the assets; you should be able to do something like this (but check my casing and spelling):

```suggestion
                                .styledBorderGradientTop,
                                .styledBorderGradientBottom
```

File: Nos/Views/Modifiers/View+StyledBorder.swift
Comment (joshuatbrown) [inline Nos/Views/Modifiers/View+StyledBorder.swift:28]: Now that we have #1375 done, we'll probably want to create a new color constant for this. Maybe this would be named something like "styled-border-shadow" in the Colors.xcassets, then you should be able to use it here like this:

```suggestion
                color: .styledBorderShadow,
```

File: NosTests/Models/NoteParserTests.swift
Comment (joshuatbrown) [reply NosTests/Models/NoteParserTests.swift:239]: Yeah, I think it's fine to just leave like this for now, and maybe ask Linda's thoughts.

File: Nos/Views/Modifiers/View+StyledBorder.swift
Comment (joshuatbrown) [inline Nos/Views/Modifiers/View+StyledBorder.swift:10]: Would you mind adding a doc comment for this? Thanks!

**Review by joshuatbrown — COMMENTED** [review]
This is looking sooooo great! I love how much better my feed looks with this! A few suggestions around documentation and SwiftGen, but otherwise this is awesome!

**Comment by martindsq** [comment]
Thanks for addressing my points @bryanmontz I think we are now closer to finish this ticket.

File: Nos/Models/NoteParser.swift
Comment (bryanmontz) [reply Nos/Models/NoteParser.swift:9]: Agree. I did that for consistency with the declaration of NoteParser right above it, but it's a different kind of thing. NoteParser is being used as a namespace here, and I think a protocol extension or caseless enum would be better for that.

**Comment by bryanmontz** [comment]
@joshuatbrown and @martindsq I've addressed all the feedback. Anything else I can do to get your approval?

I know we'll need to merge `main` but I thought we could wait to do that until it's approved to avoid multiple merges.

**Review by joshuatbrown — COMMENTED** [review]
Just one last question. This is looking awesome -- I love it!

File: Nos/Views/NoteCard.swift
Comment (bryanmontz) [reply Nos/Views/NoteCard.swift:11]: It doesn't work in all cases with `let`.

**Review by joshuatbrown — APPROVED** [review]
Excellent work!

**Review by joshuatbrown — APPROVED** [review]
Latest updates look great!

File: Nos/Models/CoreData/Event+CoreDataClass.swift
Comment (mplorentz) [reply Nos/Models/CoreData/Event+CoreDataClass.swift]: This does give me pause, because this function should only ever be called in a `context.perform` block already (but there's no way for you to know that @bryanmontz, it's just our convention) and I think I have seen nested context.perform blocks deadlock the app before. But this code has been on staging for some days and I haven't heard any reports of deadlock, so maybe it's fine? I guess I would rather it was reverted if you have the time. 

In general all the instance functions on NSManagedObjects are context agnostic and it's up to the caller to ensure that they are executed on the thread that Core Data wants. In the case of the view context it wants the main thread, and for all other contexts it wants to be in the context.perform block.

**Pull request merged**
